oshmem/mca/spml/ucx/spml_ucx.c

```diff
@@ -126,6 +126,140 @@ static ucp_request_param_t mca_spml_ucx_request_param_b = {
 };
 #endif
 
+unsigned
+mca_spml_ucx_mem_map_flags_symmetric_rkey(struct mca_spml_ucx *spml_ucx)
+{
+#if HAVE_DECL_UCP_MEM_MAP_SYMMETRIC_RKEY
+    if (spml_ucx->symmetric_rkey) {
+        return UCP_MEM_MAP_SYMMETRIC_RKEY;
+    }
+#endif
+
+    return 0;
+}
+
+int mca_spml_ucx_rkey_store_create(mca_spml_ucx_rkey_store_t *store,
+                                   ucp_worker_h worker,
+                                   int size)
+{
+    store->array = calloc(size, sizeof(*store->array));
+    if (store->array == NULL) {
+        return OSHMEM_ERR_OUT_OF_RESOURCE;
+    }
+
+    store->size   = size;
+    store->count  = 0;
+    store->worker = worker;
+    return OSHMEM_SUCCESS;
+}
+
+void mca_spml_ucx_rkey_store_destroy(mca_spml_ucx_rkey_store_t *store)
+{
+    int i;
+
+    for (i = 0; i < store->count; i++) {
+        assert(store->array[i].refcnt == 0);
+        ucp_rkey_destroy(store->array[i].rkey);
+    }
+
+    free(store->array);
+    store->array = NULL;
+    store->count = 0;
+    store->size  = 0;
+}
+
+static int mca_spml_ucx_rkey_store_find(const mca_spml_ucx_rkey_store_t *store,
+                                        const ucp_rkey_h target,
+                                        int *index)
+{
+#if HAVE_DECL_UCP_MEM_MAP_SYMMETRIC_RKEY
+    ucp_rkey_compare_params_t params = {};
+    int i, end, result, m;
+    mca_spml_ucx_rkey_t *entry;
+    ucs_status_t status;
+
+    for (i = 0, end = store->count; i < end;) {
+        m = (i + end) / 2;
+        entry = &store->array[m];
+
+        status = ucp_rkey_compare(store->worker, entry->rkey, target,
+                                  &params, &result);
+        if (status != UCS_OK) {
+            return OSHMEM_ERROR;
+        } else if (result == 0) {
+            *index = m;
+            return OSHMEM_SUCCESS;
+        } else if (result > 0) {
+            end = m;
+        } else {
+            i = m + 1;
+        }
+    }
+
+    *index = i;
+    return OSHMEM_ERR_NOT_FOUND;
+#else
+    return OSHMEM_ERROR;
+#endif
+}
+
+/* Takes ownership of input ucp remote key */
+static ucp_rkey_h mca_spml_ucx_rkey_store_get(mca_spml_ucx_rkey_store_t *store,
+                                              ucp_worker_h worker,
+                                              ucp_rkey_h target)
+{
+    int ret, i;
+
+    if ((store->size == 0) || (worker != store->worker)) {
+        return target;
+    }
+
+    ret = mca_spml_ucx_rkey_store_find(store, target, &i);
+    if ((ret == OSHMEM_ERR_NOT_FOUND) && (store->count < store->size)) {
+        memmove(&store->array[i + 1], &store->array[i],
+                (store->count - i) * sizeof(*store->array));
+        store->array[i].rkey   = target;
+        store->array[i].refcnt = 0;
+        store->count++;
```

Comment:
pls document that we insert new key to the store. maybe make it a separate function

Reply:
fixed